### PR TITLE
Use OneOrMany in the node->BoundNode map

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Internally, the compiler has a mapping from syntax nodes to potentially multiple bound nodes.  Teh multiple case is rarer, but happens with things like hidden conversions.  Because there might be multiple bound nodes associated with a syntax node, the values are stored in an ImmutableArray.  This causes a lot of unnecessary overheadgiven the common case of just a single item.  

Looking at traces in a simple typing test show 80MB for just these arrays in a simple typing test:

image

After hte change, this this gets reduced to 500k:

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@jaredpar @333fred ptal.

File: src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs [hunk 1]
```diff
@@ -139,6 +139,15 @@ public OneOrMany<TResult> Select<TResult>(Func<T, TResult> selector)
                 OneOrMany.Create(_many.SelectAsArray(selector, arg));
         }
 
+        public IEnumerable<TResult> OfType<TResult>()
+        {
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs:143]: Given how specialized this type is for single elements did you consider specializing this as well? Seems like you could make the one case a lot more effecient by doing something like: 

```csharp
public IEnumerable<TResult> OfType<TResult>()
{
  if (Count == 0) 
    return Array.Empty<TResult>();
  if (Count == 1) 
  {
    return this[0] is TResult result
      ? new TResult[] { result }
      : Array.Empty<TResult>();
  }

  return ManyCase();
  IEnumerable<TResult> ManyCase() 
  {
     foreach (var item in this)
     {
       if (item is TResult result) 
         yield return result;
     }
  }
}
```

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs:143]: So this helper is *only* used by unit tests :D.  So i didn't care.

File: src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs [hunk 2]
```diff
@@ -153,6 +154,17 @@ IEnumerator IEnumerable.GetEnumerator()
...
 
+        public OneOrMany<V> GetAsOneOrMany(K k)
+        {
+            if (_dictionary is object && _dictionary.TryGetValue(k, out var valueSet))
+            {
+                Debug.Assert(valueSet.Count >= 1);
+                return valueSet.Count == 1 ? OneOrMany.Create(valueSet[0]) : OneOrMany.Create(valueSet.Items);
+            }
+
+            return OneOrMany<V>.Empty;
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs:165]: i think GetXXXOrEmpty would make sense here.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs:165]: actually, this matches the existing method (which also returns empty in the case where you don't fine the item). 

image

File: src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs:143]: Souds good

**Review by 333fred — APPROVED** [review]
Overall LGTM, but I would suggest a minor refactoring to OfType.

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 3]
```diff
@@ -2037,7 +2037,7 @@ private Binder GetBinderToBindNode(CSharpSyntaxNode nodeToBind)
         /// Get all bounds nodes associated with a node, ordered from highest to lowest in the bound tree.
         /// Strictly speaking, the order is that of a pre-order traversal of the bound tree.
         /// </summary>
-        internal ImmutableArray<BoundNode> GetBoundNodes(CSharpSyntaxNode node)
+        internal OneOrMany<BoundNode> GetBoundNodes(CSharpSyntaxNode node)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:2040]: the nice hting is that it's all under the covers, so you could def do that.

**Comment by CyrusNajmabadi** [comment]
> Overall LGTM, but I would suggest a minor refactoring to OfType.

There's another PR that uses OneToMany.  I'll roll that suggestion into that pr to avoid havin gto rerun CI on this one :)

**Pull request merged**